src/darwin/Framework/CHIP/MTRDevice.mm

```diff
@@ -710,6 +733,130 @@ - (BOOL)_attributeDataValue:(MTRDeviceDataValueDictionary)observed satisfiesValu
+{
+    // Check whether the values coming in make sense.
+    for (MTRAttributePath * path in values) {
+        if (!MTRDataValueDictionaryIsWellFormed(values[path])) {
+            return nil;
+        }
+    }
+
+    // Check whether we have all these values already.
+    NSMutableArray<MTRAttributeRequestPath *> * requestPaths = [NSMutableArray arrayWithCapacity:values.count];
+    for (MTRAttributePath * path in values) {
+        [requestPaths addObject:[MTRAttributeRequestPath requestPathWithEndpointID:path.endpoint clusterID:path.cluster attributeID:path.attribute]];
+    }
+
+    NSArray<MTRDeviceResponseValueDictionary> * currentValues = [self readAttributePaths:requestPaths];
+
+    std::lock_guard lock(_lock);
+    auto * attributeWaiter = [[MTRAttributeValueWaiter alloc] initWithDevice:self values:values queue:queue completion:completion];
+
+    for (MTRDeviceResponseValueDictionary currentValue in currentValues) {
+        // Pretend as if this got reported, for purposes of the attribute
+        // waiter.
+        [attributeWaiter _attributeValue:currentValue[MTRDataKey] reportedForPath:currentValue[MTRAttributePathKey] byDevice:self];
+    }
+
+    if (attributeWaiter.allValuesSatisfied) {
+        MTR_LOG("%@ waitForAttributeValues no need to wait, values already match: %@", self, values);
+        [attributeWaiter _notifyWithError:nil];
+        return attributeWaiter;
+    }
+
+    // Otherwise, wait for the values to arrive or our timeout.
+    if (!self.attributeValueWaiters) {
+        self.attributeValueWaiters = [NSHashTable weakObjectsHashTable];
+    }
+    [self.attributeValueWaiters addObject:attributeWaiter];
+
+    MTR_LOG("%@ waitForAttributeValues will wait up to %f seconds for %@", self, timeout, values);
+    dispatch_source_t timerSource = dispatch_source_create(DISPATCH_SOURCE_TYPE_TIMER, 0, 0, self.queue);
+    attributeWaiter.expirationTimer = timerSource;
+
+    // Set a timer to go off after timeout, and not repeat.
+    dispatch_source_set_timer(timerSource, dispatch_time(DISPATCH_TIME_NOW, static_cast<uint64_t>(timeout * static_cast<double>(NSEC_PER_SEC))), DISPATCH_TIME_FOREVER,
+        // Allow .5 seconds of leeway; should be plenty, in practice.
+        static_cast<uint64_t>(0.5 * static_cast<double>(NSEC_PER_SEC)));
+
+    mtr_weakify(attributeWaiter);
+    mtr_weakify(self);
+    mtr_weakify(timerSource);
+    dispatch_source_set_event_handler(timerSource, ^{
+        mtr_strongify(self);
+        mtr_strongify(attributeWaiter);
+        mtr_strongify(timerSource);
+        if (self != nil && attributeWaiter != nil) {
+            [self _attributeWaitTimedOut:attributeWaiter];
+        } else if (timerSource != nil) {
+            // Make sure to cancel the timer ourselves, if we can't call _attributeWaitTimedOut.
+            dispatch_source_cancel(timerSource);
```

Comment:
> Wouldn't it be simpler for the block to hold a strong reference to the timer source and cancel it unconditionally before doing anything else?

Is it safe to dispatch_source_cancel a timer source multiple times?  If so, that would be doable, yes, since then the cancels that happen in callees of the block will just be no-ops.

Reply:
Looks like dispatch_source_cancel is in fact idempotent, so made this change.